tests/goth/test_resubscription.py

```diff
@@ -0,0 +1,189 @@
+        since the subscription identified by `subscription_id` has been created.
+        """
+        if time.time() > self.subscriptions[subscription_id] + SUBSCRIPTION_EXPIRATION_TIME:
+            logger.info("Subscription expired")
+
+            async def coro():
+                raise ApiException(
+                    http_resp=Mock(
+                        status=404,
+                        reason="Not Found",
+                        data=f"{{'message': 'Subscription [{subscription_id}] expired.'}}",
+                    )
+                )
+
+            return coro()
+        else:
+            return super().collect_offers(subscription_id, **kwargs)
+
+
+@pytest.fixture(autouse=True)
+def patch_collect_offers(monkeypatch):
+    """Install the patched `RequestorApi` class."""
+    monkeypatch.setattr(yapapi.rest.market, "RequestorApi", RequestorApi)
+
+
+async def unsubscribe_demand(sub_id: str) -> None:
+    """Auxiliary function that calls `unsubscribeDemand` operation for given `sub_id`."""
+    config = yapapi.rest.Configuration()
+    market_client = config.market()
+    requestor_api = yapapi.rest.market.RequestorApi(market_client)
+    await requestor_api.unsubscribe_demand(sub_id)
+
+
+async def assertion(events: "EventStream[Event]"):
+    """A temporal assertion that the requestor will have to satisfy."""
+
+    subscription_ids: Set[str] = set()
+
+    async def wait_for_event(event_type: Type[Event], timeout: float):
+        e = await eventually(events, lambda e: isinstance(e, event_type), timeout)
+        assert e, f"Timed out waiting for {event_type}"
+        logger.info(colors.cyan(str(e)))
+        return e
+
+    e = await wait_for_event(ComputationStarted, 10)
+
+    # Make sure new subscriptions are created at least three times
+    while len(subscription_ids) < 3:
+        e = await wait_for_event(SubscriptionCreated, SUBSCRIPTION_EXPIRATION_TIME + 10)
+        assert e.sub_id not in subscription_ids
+        subscription_ids.add(e.sub_id)
+
+    # Unsubscribe and make sure new subscription is created
+    await unsubscribe_demand(e.sub_id)
+    logger.info("Demand unsubscribed")
+    await wait_for_event(SubscriptionCreated, 5)
+
+    # Enough checking, wait until the computation finishes
+    await wait_for_event(ComputationFinished, 20)
+
+
+@pytest.mark.asyncio
+async def test_resubmission(log_dir: Path, monkeypatch) -> None:
+    """Test that checks that a demand is re-submitted after its previous submission expires."""
+
+    configure_logging(log_dir)
+
+    goth_config = load_yaml(Path(__file__).parent / "assets" / "goth-config.yml")
+
+    vm_package = await vm.repo(
+        image_hash="9a3b5d67b0b27746283cb5f287c13eab1beaa12d92a9f536b747c7ae",
+        min_mem_gib=0.5,
+        min_storage_gib=2.0,
+    )
+
+    runner = Runner(base_log_dir=log_dir, compose_config=goth_config.compose_config)
+
+    # One provider would be enough, let's not waste time for starting the second one.
```

Comment:
hmm... btw, why are we using two providers in the base test?

Reply:
Two seems to be a good default. It allows us to test more properties than would be possible with just one provider, for example that more than one agreement/worker/activity is created by the requestor.